### ISSUE TITLE
Clean up addons folder on upgrade

### ISSUE THREAD

**Issue by rkoshak** [openhab/openhab-distro#1574]
Leaving old .kar and .jar files in the addons folder after an upgrade can lead to significant problems (see https://community.openhab.org/t/endless-loop-during-start-of-oh4-0-2-service-after-upgrade-from-oh4-0-1/148774). I recommend that the installation scripts should look for and at a minimum move those files somewhere else as part of an upgrade, with logs of course. 

It gets tricky with snapshots I think because moving from one snapshot to the next within the same overall version of OH doesn't necessarily invalidate the stuff in the addons folder. But I think having OH come up in an infinite loop because the files are there is event worse.

**Comment by mstormi** [comment]
how is it supposed to work on upgrade - remove all ? Or is there a means to distinguish 'good' from 'bad' ones addons?

**Comment by holgerfriedrich** [comment]
I think minor updates and snapshots should be excluded from the proposed feature.
I would not expect losing all custom compiled add-ons after a patch level upgrade of the OH via apt upgrade.
Just my two cents.

**Comment by rkoshak** [comment]
> how is it supposed to work on upgrade - remove all ?

I would expect them all to be removed on a release upgread (e.g. 4.0 to 4.1). Testing or Snapshot releases I can imagine multiple ways, including doing nothing.

I could even see it being smart and only removing those it "knows" somehow are incompatible but to do something like that I think it would have to be implemented in the upgrade tool. 

But I think it's pretty reasonable to remove them (and by remove I mean move or even just rename so thst OH doesn't try to load them) when upgrading between release versions. In those cases the add-ons have a high likelihood of not working.

**Comment by mstormi** [comment]
you want this to be done by the postinstall script ? ping @BClark09

**Comment by Flole998** [comment]
I have a few self-built addons and if they would be deleted I would be pretty mad cause I would have to rebuild them or see where I have an original copy of those. I suggest doing at maximum a warning, no movement or deletion.

**Comment by rkoshak** [comment]
More upset than when OH won't come up or goes into an infinite loop because those add-ons are no longer compatible with OH causing it to crash (which has actually happened and why I files this issue)?

And again, not deleting, either renaming (so OH doesn't try to load them) or moving the files to a well known location. You won't lose anything.

**Comment by Flole998** [comment]
I've never had such an issue, only some of the add-ons refusing to load due to missing dependencies.

**Comment by rkoshak** [comment]
Here's another case of this problem happening.

https://community.openhab.org/t/bindings-wont-install-after-migrating-from-oh3-to-oh4/150016

**Comment by rkoshak** [comment]
With OH 4.2 release here's another case, this time with a relatively experienced user.

https://community.openhab.org/t/solved-cant-get-any-page-from-basicui-with-4-2-release/157077

At least now it doesn't seem to crash all of OH at least. But it still would be nice to at least get some logs from the upgrade scripts when it sees files in the add-ons folders. If there are not objectiosn I can submit a PR that does that much if the idea of modifying or moving the add-on files themselves remains objectionable.

**Comment by florian-h05** [comment]
I would agree on having a warning logged for each JAR/KAR inside the add-on folder during the upgrade in the postinstall script, probably before the upgrade tool runs.

**Comment by rkoshak** [comment]
Possibly related:  https://github.com/openhab/openhab-core/issues/4302

**Comment by J-N-K** [comment]
IMO whoever puts add-ons in the addons folder is responsible for ensuring that they are compatible. With the two marketplaces (community and JSON) as well as the distribution add-ons there are three supported ways of installing add-ons that ensure compatibility.

Dropping add-ons in the folder is a little bit line textual configuration: you should now what you are doing. If not, stay away from it and let the magic work.

**Comment by rkoshak** [comment]
I would tend to agree except for the fact that it becomes a lot of work helping users who experience problems because of bad jar files in the addons folder. The hardest part is even knowing the files are there. because those who have OH crash or experience OOM problems and the like never think to tell us right away that they have files in addons in the first place. At least with some warnings we can know that right away or ask whether there are warnings in the upgrade log.

In short, it's not a problem for me when upgrading with a jar file in the add-ons folder as I never have jar files there. But it is a problem for mehelping others on the forum which is why I opened the issue.
